File: app/src/main/kotlin/tech/mmarca/openvitals/data/model/HydrationReminderConfig.kt
package tech.mmarca.openvitals.data.model

import java.time.LocalTime

data class HydrationReminderConfig(
    val enabled: Boolean = false,
    val intervalMinutes: Int = DefaultIntervalMinutes,
    val activeStartTime: LocalTime = DefaultActiveStartTime,
    val activeEndTime: LocalTime = DefaultActiveEndTime,
) {
    fun normalized(): HydrationReminderConfig =
        copy(intervalMinutes = normalizeIntervalMinutes(intervalMinutes))

    companion object {
        const val DefaultIntervalMinutes = 120
        const val MinIntervalMinutes = 30
        const val MaxIntervalMinutes = 240
        const val IntervalStepMinutes = 30
        val DefaultActiveStartTime: LocalTime = LocalTime.of(7, 0)
        val DefaultActiveEndTime: LocalTime = LocalTime.of(23, 0)

        fun normalizeIntervalMinutes(value: Int): Int {
            val rounded = (value / IntervalStepMinutes) * IntervalStepMinutes
            return rounded.coerceIn(MinIntervalMinutes, MaxIntervalMinutes)
        }
    }
}

File: lib/domain/model/hydration_reminder_config.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../core/time/local_date.dart';

part 'hydration_reminder_config.freezed.dart';

@freezed
abstract class HydrationReminderConfig with _$HydrationReminderConfig {
  const HydrationReminderConfig._();

  const factory HydrationReminderConfig({
    @Default(false) bool enabled,
    @Default(HydrationReminderConfig.defaultIntervalMinutes) int intervalMinutes,
    @Default(LocalTime(7, 0)) LocalTime activeStartTime,
    @Default(LocalTime(23, 0)) LocalTime activeEndTime,
  }) = _HydrationReminderConfig;

  HydrationReminderConfig normalized() =>
      copyWith(intervalMinutes: normalizeIntervalMinutes(intervalMinutes));

  static const int defaultIntervalMinutes = 120;
  static const int minIntervalMinutes = 30;
  static const int maxIntervalMinutes = 240;
  static const int intervalStepMinutes = 30;
  static const LocalTime defaultActiveStartTime = LocalTime(7, 0);
  static const LocalTime defaultActiveEndTime = LocalTime(23, 0);

  static int normalizeIntervalMinutes(int value) {
    final rounded = (value ~/ intervalStepMinutes) * intervalStepMinutes;
    return rounded.clamp(minIntervalMinutes, maxIntervalMinutes).toInt();
  }
}
